x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/planning/RandomizedAssignmentRounding.java

```diff
@@ -0,0 +1,419 @@
+import org.elasticsearch.xpack.ml.inference.assignment.planning.AssignmentPlan.Model;
+import org.elasticsearch.xpack.ml.inference.assignment.planning.AssignmentPlan.Node;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Random;
+import java.util.Set;
+
+/**
+ * Solving allocation distribution using linear programming requires relaxing the allocation
+ * and assignment variables. This means that even though in reality they take discreet integer
+ * values, we allow the solver to provide solutions that are real numbers. This is a common
+ * technique used in LP. A common way to convert the relaxed solution back in integer values
+ * is to apply randomized rounding. This class performs randomized rounding expecting
+ * double values for the number of allocations each model gets on each node, as well as a double
+ * value in [0-1] for whether there is an assignment for each model to each node.
+ */
+class RandomizedAssignmentRounding {
+
+    private final Logger logger = LogManager.getLogger(RandomizedAssignmentRounding.class);
+
+    private final Random random;
+    private final int rounds;
+    private final Collection<Node> nodes;
+    private final Collection<Model> models;
+    private final AssignmentHolder assignmentHolder;
+
+    RandomizedAssignmentRounding(Random random, int rounds, Collection<Node> nodes, Collection<Model> models) {
+        if (rounds <= 0) {
+            throw new IllegalArgumentException("rounds must be > 0");
+        }
+        this.random = Objects.requireNonNull(random);
+        this.rounds = rounds;
+        this.nodes = Objects.requireNonNull(nodes);
+        this.models = Objects.requireNonNull(models);
+        this.assignmentHolder = new AssignmentHolder();
+    }
+
+    AssignmentPlan computePlan(Map<Tuple<Model, Node>, Double> allocationVars, Map<Tuple<Model, Node>, Double> assignmentVars) {
+        AssignmentPlan bestPlan = assignmentHolder.toPlan();
+
+        assignmentHolder.initializeAssignments(allocationVars, assignmentVars);
+        assignmentHolder.assignUnderSubscribedNodes();
+        List<Tuple<Model, Node>> softAssignmentQueue = assignmentHolder.createSoftAssignmentQueue();
+
+        if (softAssignmentQueue.isEmpty() == false) {
+            logger.debug(() -> new ParameterizedMessage("Random assignment rounding across [{}] rounds", rounds));
+            for (int i = 0; i < rounds; i++) {
+                AssignmentHolder randomizedAssignments = new AssignmentHolder(assignmentHolder);
+                randomizedAssignments.doRandomizedRounding(softAssignmentQueue);
+                AssignmentPlan randomizedPlan = randomizedAssignments.toPlan();
+                if (randomizedPlan.compareTo(bestPlan) > 0) {
+                    bestPlan = randomizedPlan;
+                }
+            }
```

Comment:
arg, I wish there was some "epsilon" we were approaching here instead of literal "randomly do things X number of times and pick the best"